### PR TITLE
Stabilize gradient clipping across trainer FT topologies

### PR DESCRIPTION
**Pull request by fzyzcjy**

### CODE REVIEW

**Comment by fzyzcjy** [comment]
/rerun-test tests/fast/utils/test_utils/test_ft_test_actions.py

**Comment by fzyzcjy** [comment]
/rerun-test tests/e2e/ft/test_trainer_ft_with_failure_dp2_cp2_real_rollout_dense.py

**Comment by fzyzcjy** [comment]
/rerun-test tests/e2e/ft/test_trainer_ft_with_failure_dp2_cp2_real_rollout_dense.py

**Comment by fzyzcjy** [comment]
> 🤖 **Posted autonomously by Claude Code acting on the user's behalf.** Summary of the bisect, the root cause, the fix on this branch, and the CI evidence for the current head; the conclusions come from nightly logs, pinned-environment reruns, and the dump/metric comparator output linked below. Please push back if any conclusion is off.

## What was failing

- `tests/e2e/ft/test_trainer_ft_with_failure_dp2_cp2_real_rollout_dense.py` passed on every main nightly from 08-18 through 08-26 and has failed on every nightly since 08-27, with identical numbers each day: the rollout-2 `train/grad_norm` gap is 6.0683% (> 5%). Last green: [run 32986112062](https://github.com/radixark/miles/actions/runs/32986112062) ([job](https://github.com/radixark/miles/actions/runs/32986112062/job/98233348873)); first red: [run 33104087365](https://github.com/radixark/miles/actions/runs/33104087365) ([job](https://github.com/radixark/miles/actions/runs/33104087365/job/98631118950)).
- Only one main commit in that window touches the training or inference path: `cd464a1c4c` Bump sglang to v0.5.18 ([#2714](https://github.com/radixark/miles/pull/2714)). Its miles-side code is not the cause: [run 33933991721](https://github.com/radixark/miles/actions/runs/33933991721) (08-27 main, old sglang, old image) passes with numbers bit-identical to the 08-26 nightly, and [run 33934001762](https://github.com/radixark/miles/actions/runs/33934001762) (08-27 main, old sglang, 08-27 image) passes as well. The only remaining variable is the `sglang-miles` source rebase onto v0.5.18, which changes the sampled content.

## Root cause

- The fault rollout (rollout 2) is not injected on main; it trains the target's own live samples (injection starts at rollout 3). That comparison is well-posed only while both sides push bit-identical weights to the engines after rollout 1.
- Normal DP2×CP2 (4-rank allreduce) and indep_dp (CP2 fold, then cross-cell fold) sum gradients in different orders. The 08-26 green run already had fp32-ulp differences in 226/452 rollout-1 grads; with the old sampled content those never flipped a bf16 weight. With the new content the rollout-1 `train/grad_norm` differs by one fp32 ulp across shardings, the active `clip_grad=1.0` (dense grad norm ≈ 1.3) multiplies that into every update, a weight flips one bf16 ulp, and temperature-0.8 sampling diverges: in the 09-03 injection experiment ([job](https://github.com/radixark/miles/actions/runs/33696774224/job/100469173373)) the target's own rollout-2 samples match the baseline recording at only 0.9265 mean token match, and forcing identical rollout-2 data drops the gap from 6.0683% to 0.0180%.

## Fix on this branch (base pinned to `1e3b7a0887`, the 09-02 nightly main)

- In `dp2_cp2_real_rollout_dense`, both baseline and target now run with `--debug-deterministic-collective` (same fixed-tree fold for both reduction topologies, as `scenario_deterministic` already does) and `--clip-grad 10.0` (clipping inactive; `train/grad_norm` is still computed and compared). Main's rollout-3 injection is kept unchanged; the injection start does not move.
- No threshold, skip/allow pattern, sample count, fault action, or topology changes. Net diff vs base: `scenario_with_failure.py` +3, `README.md` +1, new `tests/fast/e2e/ft/test_scenario_with_failure.py` pinning the contract ([diff](https://github.com/radixark/miles/compare/1e3b7a0887...f5db4fb5380ee1f35987cd990459137c5e45eb85)).

## Evidence on this branch

Environment pinned to the 09-02 nightly: image `dev-202609021245`, sglang-miles `98bb1455b9`, Megatron `8c1e05747`.

| Check | Run | Result |
| --- | --- | --- |
| H200 dense test (`b0d999afb7`; scenario unchanged since) | [run 34093950350](https://github.com/radixark/miles/actions/runs/34093950350) / [job](https://github.com/radixark/miles/actions/runs/34093950350/job/101653472953) | `With-failure comparison test PASSED`; grad_norm gap rollout 1 = 7.15e-07, rollout 2 = 0 (bit-identical), rollout 3 = 0.39%; rollout-1 params and grads bit-identical (452/452 each); dumps rollout 1/2/3 = 0 failed; shrink + heal events present; injection guard 0.5969; 24 min 54 s |
| Fast contract test (`f5db4fb538`) | [run 34099313396](https://github.com/radixark/miles/actions/runs/34099313396) | `3 passed` |
| Regular PR CI (`f5db4fb538`) | [checks](https://github.com/radixark/miles/pull/3090/checks) | 17 success / 8 skipped / 0 red |
| Earlier H200 attempt (`b0d999afb7`) | [run 34092352012](https://github.com/radixark/miles/actions/runs/34092352012) | Failed at target phase-B startup with `EADDRINUSE port 20988` before any comparison (free-port race, infra flake); rerun above passed |

## Not covered here

- Since the 09-04 trainer-FT refactor, main's other with_failure and deterministic FT variants fail on `CellReconfigureEvent sequence mismatch` (missing heal event after stop/start) and the post-crash survivor wait grew from ~125 s to ~685 s (e.g. [run 34041343321](https://github.com/radixark/miles/actions/runs/34041343321)). Both are independent of this PR; this branch is intentionally based on the pre-refactor main and should be re-validated after those are fixed on main.
- Clipping is no longer exercised by this test; the other FT variants already run with grad norms below 1.0, so this brings the dense mode in line with them.

**Comment by fzyzcjy** [comment]
put into deliver-2 chain
